### PR TITLE
[Inductor] Fix log_sigmoid_forward flushing float32 subnormals to +0.0 on CUDA

### LINKED ISSUE #188541: [Inductor][CUDA] F.logsigmoid flushes negative float32 subnormal outputs to +0.0, changing sign-sensitive behavior
### 🐛 Describe the bug

When compiling torch.nn.functional.logsigmoid with the Inductor backend on CUDA, large positive float32 inputs produce different results from eager mode.

For inputs where eager mode returns negative nonzero subnormal values, the compiled function returns +0.0. This also changes downstream observable behavior such as torch.signbit, equality-to-zero checks, and torch.copysign.
### Reproduce script
```python
import torch
import torch.nn.functional as F

def fn(x):
    y = F.logsigmoid(x)
    return y, torch.signbit(y), y == 0, torch.copysign(torch.ones_like(y), y)

cfn = torch.compile(fn, backend="inductor", fullgraph=True)

xs = torch.tensor([80.0, 85.0, 87.0, 88.0, 90.0, 100.0],
                  device="cuda", dtype=torch.float32)

eager = fn(xs)
compiled = cfn(xs)

print("x:", xs)
print("eager y:", eager[0])
print("compiled y:", compiled[0])
print("eager signbit:", eager[1])
print("compiled signbit:", compiled[1])
print("eager y == 0:", eager[2])
print("compiled y == 0:", compiled[2])
print("eager copysign:", eager[3])
print("compiled copysign:", compiled[3])
```

### Output
```text
x: tensor([ 80.,  85.,  87.,  88.,  90., 100.], device='cuda:0')
eager y: tensor([-1.8049e-35, -1.2161e-37, -1.6458e-38, -6.0546e-39, -8.1940e-40,
        -3.7835e-44], device='cuda:0')
compiled y: tensor([-1.8049e-35, -1.2161e-37, -1.6458e-38,  0.0000e+00,  0.0000e+00,
         0.0000e+00], device='cuda:0')
eager signbit: tensor([True, True, True, True, True, True], device='cuda:0')
compiled signbit: tensor([ True,  True,  True, False, False, False], device='cuda:0')
eager y == 0: tensor([False, False, False, False, False, False], device='cuda:0')
compiled y == 0: tensor([False, False, False,  True,  True,  True], device='cuda:0')
eager copysign: tensor([-1., -1., -1., -1., -1., -1.], device='cuda:0')
compiled copysign: tensor([-1., -1., -1.,  1.,  1.,  1.], device='cuda:0')
```

### Versions

PyTorch: 2.14.0.dev20260624+cu130
CUDA used to build PyTorch: 13.0
OS: Ubuntu 24.04.4 LTS (x86_64)
Is CUDA available: True
CUDA runtime version: 12.0.140

cc @ezyang @gchanan @kadeng @msaroufim @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @ptrblck @eqy @tinglvv @nWEIdia @chauhang @penguinwu @voznesenskym @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @muchulee8 @amjames @aakhundov @coconutruben @jataylo

### PR DESCRIPTION
**Pull request by heytanix**
## Summary

Fixes #188541

- **Root cause**: Triton kernels run with FTZ enabled by default (`disable_ftz: False` in `triton_meta`). NVIDIA libdevice does NOT call `log1pf` as an opaque intrinsic — LLVM inlines it as a polynomial of `fma.rn.ftz.f32` PTX instructions. For subnormal inputs, the final reconstructing FMA in that polynomial is itself FTZ-tagged and flushes the correctly-computed-but-subnormal result to zero. This is confirmed from actual PTX dump of the compiled kernel (not inferred from I/O behavior). Note that `exp()`'s subnormal output is **not** flushed — `libdevice.expf`'s final step uses `mul.f32` without `.ftz`. The loss is specific to `log1p`'s output. `log(1+x)` fails too, but for an unrelated reason (float32 cancellation: `1.0f + 6e-39f = 1.0f` exactly due to precision, not FTZ).

- **Effect**: For float32 `x >= ~87.3`, `logsigmoid(x) ~= -exp(-x)` is a subnormal (~6e-39 for x=88). The compiled path returns `0.0` instead of the correct negative subnormal, silently flipping `torch.signbit`, `y == 0`, and `torch.copysign` comparisons. This corrupts training loss when `BCEWithLogitsLoss` / `LogSigmoid` is used with large logits.

- **Fix**: Add `aten.log_sigmoid_forward` to `decomps_to_exclude` and register a device-conditional `@register_lowering`:
  - **CUDA**: `FallbackKernel` to the native ATen kernel (nvcc, no `-ftz`). CUDA backward ignores the buffer (confirmed in `Activation.cpp:799`), so the empty buffer from `FallbackKernel` is correct and introduces no backward regression.
  - **CPU/other backends**: Inline decomposition preserved for fusion. These backends have no FTZ issue; standard-library `log1p` handles subnormals correctly. Buffer semantics match upstream: XPU gets `empty([0])`, CPU and others get `z = exp(-|x|)` as the buffer (matching what `log_sigmoid_backward_cpu` reads back).
  - A global `make_fallback` was rejected because it would regress CPU fusion with no correctness benefit.

## Before / After (RTX 4070 Laptop, CUDA 12.6, float32)

```
x = [88.0, 90.0, 100.0]

# Before fix (compiled):
logsigmoid: [0.0, 0.0, 0.0]       ← WRONG
signbit:    [False, False, False]  ← WRONG (all should be True)

# After fix (compiled = eager):
logsigmoid: [-6.055e-39, -8.194e-40, -3.784e-44]  ✓
signbit:    [True, True, True]                      ✓
```

## Test plan

Two new regression tests added to `test/inductor/test_cuda_repro.py`:

```bash
# Forward: bitwise match on values + signbit + ==0
python test/inductor/test_cuda_repro.py -k test_logsigmoid_subnormal_outputs -v

# Backward: gradients match at large x (subnormal gradient range), float64
python test/inductor/test_cuda_repro.py -k test_logsigmoid_subnormal_backward -v
```

Float32 backward was also verified explicitly: the backward decomposition uses `z/(1+z)` (not `log1p`), so subnormal gradients are not FTZ-flushed and are correct without a backward fix. Separately confirmed: `gradcheck` passes on both devices, `nn.LogSigmoid` / `BCEWithLogitsLoss` / mixed-range tensors all match eager exactly.

🤖 Authored with [Claude Code](https://claude.ai/claude-code)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @jataylo

### CODE REVIEW

**Comment by jansel** [comment]
@claude explain those CI failures, check if they are related to this PR, and review the changes

**Review by jansel — CHANGES_REQUESTED** [review]
address automated feedback from claude

**Comment by heytanix** [comment]
Addressed — gated the CUDA fallback on `dtype == torch.float32` (commit 372bd23). float64 and float16/bfloat16 now keep the fused inline decomposition.

**Review by jansel — CHANGES_REQUESTED** [review]
Can this be fixed without falling back?

**Comment by heytanix** [comment]
No fallback needed - replaced the `FallbackKernel` with an inline `where` branch in the lowering.

For `x > 87.3` (where `exp(-x)` becomes subnormal), we compute `-exp(-x)` directly instead of `min(0,x) - log1p(exp(-|x|))`. At those magnitudes `log1p(z) = z` exactly in float32 (the next term `z^2/2` is far below the float32 minimum subnormal), so there is no precision loss and the op stays fused.

For all other inputs the standard decomposition runs unchanged.

CPU tests pass including the subnormal range (x=88, 90, 100 all match eager bit-for-bit). CUDA verification will come from CI.

**Review by jansel — APPROVED** [review]

**Review by eellison — COMMENTED** [review]
I'm a bit concerned about the proliferation of potentially expensive subnormal/-0 prs. extra instructions are not always free. can we config gate them ?

**Comment by heytanix** [comment]
@eellison Benchmarked the overhead of the `where` branch on an RTX 4070 Laptop GPU (CUDA 12.4):

```
N=2^20  ( 4 MB)  baseline= 12.5 us  fixed= 12.5 us  overhead= -0.0%
N=2^22  (16 MB)  baseline= 35.2 us  fixed= 34.3 us  overhead= -2.3%
N=2^24  (64 MB)  baseline=594.0 us  fixed=594.7 us  overhead= +0.1%
```

The overhead is zero - the kernel is memory-bandwidth-bound at all practical sizes, so the extra `gt`, `exp`, and `where` instructions are fully hidden behind the load/store latency. No config gate needed.

**Review by eellison — COMMENTED** [review]
Yes, it's memory bound here, but potentially other kernels it won't be.

**Comment by heytanix** [comment]
@eellison  to clarify - the current fix is an inline `where` branch in Triton, so there are extra `gt`, `exp`, and `where` instructions.
Benchmarked it on my local RTX 4070 and the overhead is zero since the kernel is memory-bound (compute hides behind load/store). Happy to add a config gate if you want that pattern established for future PRs though.
